chainstate/src/detail/chainstateref/epoch_seal.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_storage::{BlockchainStorageWrite, SealedStorageTag};
+use common::{
+    chain::{ChainConfig, GenBlockId},
+    primitives::BlockHeight,
+};
+use pos_accounting::{FlushablePoSAccountingView, PoSAccountingDB, PoSAccountingDeltaData};
+use utils::tap_error_log::LogError;
+
+use crate::BlockError;
+
+pub enum BlockStateEvent {
+    Connect(BlockHeight),
+    Disconnect(BlockHeight),
+}
+
+pub fn activate_epoch_seal<S: BlockchainStorageWrite>(
+    db_tx: &mut S,
+    chain_config: &ChainConfig,
+    block_op: BlockStateEvent,
+) -> Result<(), BlockError> {
+    match block_op {
+        BlockStateEvent::Connect(tip_height) => {
+            if chain_config.is_due_for_epoch_seal(&tip_height) {
+                advance_epoch_seal(db_tx, chain_config, tip_height)?;
+            }
+        }
+        BlockStateEvent::Disconnect(tip_height) => {
+            let disconnected_tip = tip_height.next_height();
+            if chain_config.is_due_for_epoch_seal(&disconnected_tip) {
+                rollback_epoch_seal(db_tx, chain_config, disconnected_tip)?;
+            }
+        }
+    };
+    Ok(())
+}
+
+fn advance_epoch_seal<S: BlockchainStorageWrite>(
+    db_tx: &mut S,
+    chain_config: &ChainConfig,
+    tip_height: BlockHeight,
+) -> Result<(), BlockError> {
+    let current_epoch_index = chain_config.epoch_index_from_height(&tip_height);
+    let epoch_index_to_seal =
+        current_epoch_index - chain_config.sealed_epoch_distance_from_tip() as u64;
+    let epoch_length = chain_config.epoch_length().get();
+    let first_block_epoch_to_seal = epoch_index_to_seal * epoch_length;
+    let end_block_epoch_to_seal = first_block_epoch_to_seal + epoch_length;
+
+    // iterate over every block in the epoch and merge every block delta into a singe delta
+    let epoch_delta = (first_block_epoch_to_seal..end_block_epoch_to_seal)
+        .try_fold(
```

Comment:
Well, if we're going through all this work to create the delta and we don't want to delete it, doesn't it make sense to just not recreate this delta unless we can't find it in the database?

Why are we not storing deltas from epochs in the database?

Reply:
Then please go for it. Only the merge should be done at the epoch threshold.